File: web/src/lib/api/menu.ts
import { http } from '@/lib/http'
import type { Menu, MenuNode } from '@/lib/api/types'

export interface CreateMenuPayload {
  parent_id: string
  name: string
  type: number
  path?: string | null
  component?: string | null
  perm?: string | null
  api_path?: string | null
  api_method?: string | null
  icon?: string | null
  sort?: number
  visible?: number
  status?: number
}

export type UpdateMenuPayload = Partial<CreateMenuPayload>

export const menuApi = {
  list: () => http.get<MenuNode[]>('/menus'),
  detail: (id: string) => http.get<Menu>(`/menus/${id}`),
  create: (payload: CreateMenuPayload) => http.post<Menu>('/menus', payload),
  update: (id: string, payload: UpdateMenuPayload) =>
    http.put<Menu>(`/menus/${id}`, payload),
  remove: (id: string) => http.delete<null>(`/menus/${id}`),
}

File: crates/service/src/dto/menu.rs
use serde::{Deserialize, Serialize};
use validator::Validate;

#[derive(Debug, Deserialize, Validate, utoipa::ToSchema)]
pub struct CreateMenuReq {
    #[serde(default, with = "entity::id::string")]
    #[schema(value_type = String)]
    pub parent_id: i64,
    #[validate(length(min = 1, max = 64))]
    pub name: String,
    pub r#type: i16,
    pub path: Option<String>,
    pub component: Option<String>,
    pub perm: Option<String>,
    pub api_path: Option<String>,
    pub api_method: Option<String>,
    pub icon: Option<String>,
    pub sort: Option<i32>,
    pub visible: Option<i16>,
    pub status: Option<i16>,
}

#[derive(Debug, Deserialize, Validate, utoipa::ToSchema)]
pub struct UpdateMenuReq {
    #[serde(with = "entity::id::string_opt", default)]
    #[schema(value_type = Option<String>)]
    pub parent_id: Option<i64>,
    #[validate(length(min = 1, max = 64))]
    pub name: Option<String>,
    pub r#type: Option<i16>,
    pub path: Option<String>,
    pub component: Option<String>,
    pub perm: Option<String>,
    pub api_path: Option<String>,
    pub api_method: Option<String>,
    pub icon: Option<Option<String>>,
    pub sort: Option<i32>,
    pub visible: Option<i16>,
    pub status: Option<i16>,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct MenuNode {
    #[serde(flatten)]
    pub menu: entity::menu::Model,
    #[schema(no_recursion)]
    pub children: Vec<MenuNode>,
}
